delegation/src/Pos/Delegation/Cede/Logic.hs

```diff
@@ -86,73 +89,61 @@ detectCycleOnAddition toAdd
-    => PublicKey                             -- ^ Issuer
-    -> PublicKey                             -- ^ Delegate
-    -> ProxySKHeavy                          -- ^ i->d psk
-    -> m Bool
-dlgReachesIssuance i d _ | i == d = pure True
-dlgReachesIssuance i d psk = reach i
+    => StakeholderId          -- ^ Issuer
+    -> m (Maybe ProxySKHeavy) -- ^ Last cert in the chain, if any
+dlgLastPsk i = reach (Nothing, i)
   where
-    -- Delegate 'd' has right to issue block instead of issuer 'i' if
-    -- there's a delegation chain:
-    --
-    -- i → x₁ → x₂ → … xₖ → d
-    --
-    -- where every arrow is resolved psk, and the last one xₖ → d
-    -- equals to the passed one.
-    reach curUser = getPsk (addressHash curUser) >>= \case
-        Nothing   -> pure False
-        Just psk'
-            | pskDelegatePk psk' == d -> pure $ psk' == psk
-            | otherwise               -> reach (pskDelegatePk psk')
+    reach (!prevPsk, !curUser) = getPsk curUser >>= \case
+        Nothing    -> pure prevPsk
+        Just !psk' -> reach (Just psk', addressHash $ pskDelegatePk psk')
 
 -- | Verifies a header from delegation perspective (signature checks).
 dlgVerifyHeader ::
        (MonadCedeRead m)
     => MainBlockHeader
     -> m (Either Text ())
 dlgVerifyHeader h = runExceptT $ do
-    -- Issuer didn't delegate the right to issue to elseone.
-    let issuer = h ^. mainHeaderLeaderKey
+    let bIssuer = h ^. mainHeaderLeaderKey
     let sig = h ^. gbhConsensus . mcdSignature
-    issuerPsk <- getPskPk issuer
-    whenJust issuerPsk $ \psk -> case sig of
-        (BlockSignature _) ->
-            throwError $
-            sformat ("issuer "%build%" has delegated issuance right, "%
-                     "so he can't issue the block, psk: "%build%", sig: "%build)
-                issuer psk sig
-        _ -> pass
-
-    -- Check that if proxy sig is used, delegate indeed has right to
-    -- issue the block. Signatures themselves are checked in the
-    -- constructor, here we only verify they are related to slot
-    -- leader. Self-signed proxySigs are forbidden on block
-    -- construction level.
-    case h ^. gbhConsensus ^. mcdSignature of
+
+    case sig of
+        -- If slot leader delegated, nobody can issue blocks with default
+        -- signatures and his name in "slot leader" field.
+        (BlockSignature _) -> do
+            whenJustM (getPskPk bIssuer) $ \psk ->
+                throwError $
+                sformat ("issuer "%build%" has delegated issuance right, "%
+                         "so he can't issue the block himself, psk: "%build%", sig: "%build)
+                    bIssuer psk sig
+        -- We check that the cert is related to block issuer. This
+        -- also checks that the delegate didn't delegate himself.
         (BlockPSignatureHeavy pSig) -> do
-            let psk = psigPsk pSig
-            let delegate = pskDelegatePk psk
-            canIssue <- dlgReachesIssuance issuer delegate psk
```

Comment:
Hm, did the old version check that `delegate` didn't re-delegate anything further?